### ISSUE TITLE
failure at Qed time but only with primitive projections

### ISSUE THREAD

**Issue by jesboat** [coq/coq#19082]
### Description of the problem

I have a weird set of records (defined with `Inductive`), and a definition defined by a proof script with mostly `refine`. When no records are declared with primitive projections, the proof script goes through ok. However, when one (specific) one is declared with primitive projections, the tactics generate an invalid proof which fails at Qed time.

There's an example attached.

The precise type error is

```
Illegal application: 
The term "OutputWrapper" of type
 "forall bob ow_alice : Type, output_inner ow_alice bob -> output_wrapper bob"
cannot be applied to the terms
 "next_bob" : "Type"
 "req0" : "request orig_bob"
The 2nd term has type "request orig_bob" which should be a subtype of
 "Type".
Raised at Exninfo.iraise in file "clib/exninfo.ml", line 79, characters 4-11
Called from Typing.execute in file "pretyping/typing.ml", line 562, characters 24-44
``` 

### Small Coq file to reproduce the bug

```coq
Inductive output_inner (alice bob : Type) := Common { }.

Inductive output_wrapper (bob : Type) := OutputWrapper {
    ow_alice : Type;
    ow_inner  : output_inner ow_alice bob;
}.

#[projections(primitive)]
Inductive request (orig_bob : Type) := InputWrapper {
      req_alice : Type;
      req_next_bob := orig_bob
}.

Definition do_the_thing
              [orig_bob]
              (req : request orig_bob)
          : output_wrapper orig_bob.
Proof.
  refine (
      match req with
      | {|
          req_alice := alice;
          req_next_bob := next_bob;
       |} => _
      end
  ).
  Validate Proof.
  refine (
    let wip := @OutputWrapper next_bob alice
    in _
  ).
  (* Validate Proof.   <--- fails when primitive projections *)
  apply wip.
  constructor.
Qed.
```

### Version of Coq where this bug occurs

8.19.0, 8.19.1

**Comment by SkySkimmer** [comment]
Probably some kind of debruijn bug
8.11 seems fine, will bisect later

**Comment by Janno** [comment]
I used the simplified example to bisect this:
```
f7de37ebda9debd9f61a2b9f3c153522fb0ffa55 is the first bad commit
commit f7de37ebda9debd9f61a2b9f3c153522fb0ffa55
Author: Hugo Herbelin <Hugo.Herbelin@inria.fr>
Date:   Thu Jan 27 18:37:53 2022 +0100

    Replacing a decompose_lam_n_assum which should have been a decompose_lam_n_decls.
    
    Clarifying that the other use is ok.

 pretyping/inductiveops.ml | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)
```
cc @herbelin

**Comment by herbelin** [comment]
Thanks for the bisection and the reduced example. Fixed in #19088.

**Comment by herbelin** [comment]
Thanks for your support!
